libc/src/ringbuffer.c

```diff
@@ -0,0 +1,97 @@
+#include <ringbuffer.h>
+#include <stdlib.h>
+
+int ringbuffer_init(ringbuffer_t* ref, size_t sz) {
+    ref->sz = sz;
+    ref->unr_data = 0;
+
+// alloc the needed size for this ringbuff and use kmalloc
+// if this is compiled with the proper flag
+#ifndef _KERNEL_
+    ref->data = (uint8_t*) malloc(sz);
+#else
+    ref->data = (uint8_t*) kmalloc(sz);
+#endif
+
+    if (ref->data == NULL) {
+        return -1;
+    }
+
+    ref->r_pos = 0;
+    ref->w_pos = 0;
+
+    return 0;
+}
+
+ringbuffer_t* ringbuffer_new(size_t sz) {
+    ringbuffer_t* ref = malloc(sizeof(ringbuffer_t));
+    if (ringbuffer_init(ref, sz) == -1) {
+#ifndef _KERNEL_
+        free(ref);
+#else
+        kfree(ref);
+#endif
+        ref = NULL;
+    }
+
+    return ref;
+}
+
+size_t ringbuffer_used(ringbuffer_t* ref) {
+    return ref->unr_data;
+}
+
+size_t ringbuffer_free(ringbuffer_t* ref) {
```

Comment:
The naming is a bit unfortunate here. I'd suggest `ringbuffer_available_space`, and renaming `ringbuffer_dispose` to `ringbuffer_free`.